compiler/rustc_typeck/src/check/pat.rs

```diff
@@ -2032,12 +2032,36 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         {
             if let (Some(span), true) = (ti.span, ti.origin_expr) {
                 if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
-                    err.span_suggestion(
-                        span,
-                        "consider slicing here",
-                        format!("{}[..]", snippet),
-                        Applicability::MachineApplicable,
-                    );
+                    let applicability = match self.resolve_vars_if_possible(ti.expected).kind() {
+                        ty::Adt(adt_def, _)
+                            if self.tcx.is_diagnostic_item(sym::Option, adt_def.did)
+                                || self.tcx.is_diagnostic_item(sym::Result, adt_def.did) =>
+                        {
+                            // Slicing won't work here, but `.as_deref()` might (issue #91328).
+                            err.span_suggestion(
+                                span,
+                                "consider using `as_deref` here",
+                                format!("{}.as_deref()", snippet),
+                                Applicability::MaybeIncorrect,
+                            );
+                            None
+                        }
+                        ty::Adt(adt_def, _)
+                            if self.tcx.is_diagnostic_item(sym::Vec, adt_def.did) =>
```

Comment:
So...I think I'm learning something today about how Rust's `SliceIndex` works (at least, unless I've missed something). So...I guess when we do `a_vec[..]`, what's *actually* happening is there is an implicit deref: `(a_vec.deref())[..])` - because `SliceIndex` is only implemented for `slice`, then the implicit deref is basically "forced". (At least, I think this is the case...please someone correct me if I'm wrong.)

I bring this up because rather than checking for `Vec` exactly, we could check if the type impls `Deref<Target=X>`?

Of course, this could be done as followup. This looks good otherwise.

Reply:
I guess we could in general be smarter above too for `Result<T>` to only suggest if `T: Deref<Target=X>`

Reply:
Yes, that would be nicer, but I don't know how to check for such a trait bound here. Let's leave it to a follow-up (or do you know an easy way to implement this?).

Reply:
Yes, let's leave this as a followup. Can you add a FIXME comment though?